src/Microsoft.Azure.WebJobs.Extensions.Kafka/Listeners/KafkaTopicScaler.cs

```diff
@@ -0,0 +1,247 @@
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.OfType<KafkaTriggerMetrics>().ToArray());
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext<KafkaTriggerMetrics> context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.ToArray());
+        }
+
+        private ScaleStatus GetScaleStatusCore(int workerCount, KafkaTriggerMetrics[] metrics)
+        {
+            var status = new ScaleStatus
+            {
+                Vote = ScaleVote.None,
+            };
+
+            const int NumberOfSamplesToConsider = 5;
+
+            // At least 5 samples are required to make a scale decision for the rest of the checks.
+            if (metrics == null || metrics.Length < NumberOfSamplesToConsider)
+            {
+                return status;
+            }
+
+            var lastMetrics = metrics.Last();
+            long totalLag = lastMetrics.TotalLag;
+            long partitionCount = lastMetrics.PartitionCount;
+            long lagThreshold = 1000L;
+
+            //This case shouldn't ever happens as we do not scale out in case workerCount >= partitionCount
+            if (partitionCount > 0 && partitionCount < workerCount)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+
+                if (this.logger.IsEnabled(LogLevel.Information))
+                {
+                    this.logger.LogInformation("WorkerCount ({workerCount}) > PartitionCount ({partitionCount}). For topic {topicName}, for consumer group {consumerGroup}.", workerCount, partitionCount, this.topicName, this.consumerGroup);
+                    this.logger.LogInformation("Number of instances ({workerCount}) is too high relative to number of partitions ({partitionCount}). For topic {topicName}, for consumer group {consumerGroup}.", workerCount, partitionCount, this.topicName, this.consumerGroup);
+                }
+
+                return status;
+            }
+
+
+            // Check to see if the Kafka consumer has been empty for a while. Only if all metrics samples are empty do we scale down.
+            bool partitionIsIdle = metrics.All(p => p.TotalLag == 0);
+
+            if (partitionIsIdle)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+                if (this.logger.IsEnabled(LogLevel.Information))
+                {
+                    this.logger.LogInformation("Topic '{topicName}', for consumer group {consumerGroup}' is idle.", this.topicName, this.consumerGroup);
+                }
+
+                return status;
+            }
+
+
+            // don't scale out beyond the number of partitions, Maintain a minimum ratio of 1 worker per 1,000 unprocessed messages.
+            if (metrics[0].TotalLag > 0 && totalLag / lagThreshold > partitionCount)
```

Comment:
Can you separate the worker ration check from the queue length increasing check as in EventHubs [here](https://github.com/Azure/azure-functions-eventhubs-extension/blob/daab3ae8a8e08dd971b8f9b19d52f09ef44af081/src/Microsoft.Azure.WebJobs.Extensions.EventHubs/Listeners/EventHubsScaleMonitor.cs#L323) to make the checks distinct/clear?

Reply:
The Eventhub code may cause the vote to scale out regardless about the number of partitions then the we scale it down again to not exceed the number of partitions . 

Also it will scale out just based on the last value, should we at least check the last few values to check for increase?